### PR TITLE
RATIS-1175. DataStreamOutput should support FileChannel#transferTo.

### PR DESCRIPTION
**Pull request by szetszwo**
See https://issues.apache.org/jira/browse/RATIS-1175

### CODE REVIEW

File: ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java [hunk 1]
```diff
@@ -22,25 +22,39 @@
...
 import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
 import java.util.concurrent.CompletableFuture;
 
 /**
  * Streaming client implementation
  * allows client to create streams and send asynchronously.
  */
 public class DataStreamClientImpl implements DataStreamClient {
+  private static RaftClientReply getRaftClientReply(DataStreamReply reply) {
```
Comment (runzhiwang) [inline ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java:49]: Duplicate with the following code.
```
static RaftClientReply getRaftClientReply(DataStreamReply dataStreamReply) {
    if (dataStreamReply instanceof DataStreamReplyByteBuffer) {
      try {
        return ClientProtoUtils.toRaftClientReply(((DataStreamReplyByteBuffer) dataStreamReply).slice());
      } catch (InvalidProtocolBufferException e) {
        throw new IllegalStateException("Failed to parse " + JavaUtils.getClassSimpleName(dataStreamReply.getClass())
            + ": reply is " + dataStreamReply, e);
      }
    } else {
      throw new IllegalStateException("Unexpected " + dataStreamReply.getClass() + ": reply is " + dataStreamReply);
    }
  }
```

**Comment by runzhiwang** [comment]
@szetszwo Hi, in transferTo method there are three methods: transferToDirectly which use sendfile and is the fastest, transferToTrustedChannel which use MapByteBuffer, transferToArbitraryChannel which use HeapByteBuffer and is the slowest. Our code use the slowest transferToArbitraryChannel , which will use HeapByteBuffer as the image shows. Not sure whether can we support transferToDirectly or transferToTrustedChannel.

image

File: ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java (hunk 1 above)
Comment (szetszwo) [reply ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java:49]: Thanks.  Will move them to ClientProtoUtils.

**Comment by szetszwo** [comment]
@runzhiwang , good observation.   I guess it depends on the ByteBuffer type in the source.  If it is direct buffer, it will use transferToDirectly.  The test uses FileInputStream which probably is not using direct buffer.

**Comment by szetszwo** [comment]
@runzhiwang ,  just found that our WritableByteChannel should at least extend SelectableChannel.  Then, it may be able to use transferToDirectly.  This needs some research.

How about we commit this first and continue the work in another JIRA?

**Comment by runzhiwang** [comment]
@szetszwo Sure, thanks for explanation.

**Review by runzhiwang — APPROVED** [review]
LGTM, waiting CI pass

**Comment by szetszwo** [comment]
Another way is to provide a different API for transferring files, say
```
  CompletableFuture<DataStreamReply> writeAsync(RandomAccessFile);
```
Then, we may use Netty DefaultFileRegion for zero-copy file transfer.
https://github.com/netty/netty/blob/4.1/example/src/main/java/io/netty/example/file/FileServerHandler.java#L53

**Comment by runzhiwang** [comment]
@szetszwo  I agree. Do you think it can be used in ozone case, maybe can be used to write file smaller than 128MB, i.e. the size of block ? For file bigger than 128MB, we must split it into multi blocks, and can not use transferTo.

**Comment by runzhiwang** [comment]
Sorry my bad, transferTo support offset, for big file, we can also use transferTo to write multi blocks.

**Comment by szetszwo** [comment]
We still can use transferTo for file > 128MB since the API can specify position and count.
https://netty.io/4.1/api/io/netty/channel/DefaultFileRegion.html#DefaultFileRegion-java.nio.channels.FileChannel-long-long-

User may pass a large file.  Internally, we send it by chunks.

**Comment by runzhiwang** [comment]
transferTo can avoid reading data to user space, it transfer data between kernel space and socket/disk. 

But we still need to calculate checksum, so client still need read data to user space. And if we use transferTo, client transfer data to three nodes in parallel maybe best, otherwise if we use client -> primary -> peer, primary must wait finish write 128MB block, and then transfer the block file to follower, it maybe slow, but if client transfer data to three nodes in parallel, "the client has a higher network bandwidth requirement and it cannot optimise the network traffic". So not sure whether can use transferTo in ozone.

**Comment by szetszwo** [comment]
> ... , otherwise if we use client -> primary -> peer, primary must wait finish write 128MB block, and then transfer the whole block file to follower ...

No.  The transferTo method calls WritableByteChannel.write(..) multiple times.  For example, transferToArbitraryChannel(..) create a 8KB buffer and call WritableByteChannel.write(..) every 8KB.  We must send small packets in Streaming.
```
//sun.nio.ch.FileChannelImpl
  private long transferToArbitraryChannel(long var1, int var3, WritableByteChannel var4) throws IOException {
    int var5 = Math.min(var3, 8192);
    ByteBuffer var6 = Util.getTemporaryDirectBuffer(var5);
```
var6 above is the 8KB buffer.

**Comment by szetszwo** [comment]
> But we need to calculate checksum, so client still need read data, to avoid reading data to user space, we can use MapByteBuffer. so we need 2 times read data from disk to kernel space: transferTo and MapByteBuffer.

We may pass MapByteBuffer to our writeAsync(..) method.  We should benchmark it.  I suspect it is better than transferTo(..) since it is using transferToArbitraryChannel(..).

**Comment by szetszwo** [comment]
@runzhiwang , thanks a lot for reviewing this.  Let's continue the discussion on https://issues.apache.org/jira/browse/RATIS-1176

**Pull request merged**